writeimage.sh

```diff
@@ -78,31 +76,43 @@ if ! [ -f "$DISK_IMG" ]; then
     exit 1
 fi
 
-gunzip=$(which unpigz 2> /dev/null || which gunzip 2> /dev/null || true)
-unxz=$(which unxz 2> /dev/null || true)
+if [[ $(id -u) -ne 0 ]]; then echo "please run as root"; exit 1; fi
+
+gzip=$(which pigz 2>/dev/null || which gzip 2>/dev/null || true)
+xz=$(which xz 2>/dev/null || true)
+pv=$(which pv 2>/dev/null || true)
+ddopts="status=none"
+if [ -z "$pv" ]; then
+	pv=cat
+	ddopts="status=progress"
+fi
 
+extractor=""
+extractopts=(-d -c)
 if [[ $DISK_IMG == *.gz ]]; then
-    if [ -z "$gunzip" ]; then
+    if [ -z "$gzip" ]; then
         msg "make sure you have the gzip package installed"
         exit 1
     fi
-    msg "decompressing the .gz compressed image"
-    $gunzip -c "$DISK_IMG" > "${DISK_IMG%???}"
-    DISK_IMG=${DISK_IMG%???}
+    extractor=$gzip
 elif [[ $DISK_IMG == *.xz ]]; then
-    if [ -z "$unxz" ]; then
+    if [ -z "$xz" ]; then
         msg "make sure you have the xz package installed"
         exit 1
     fi
-    msg "decompressing the .xz compressed image"
-    $unxz -T 0 -c "$DISK_IMG" > "${DISK_IMG%???}"
-    DISK_IMG=${DISK_IMG%???}
+	extractor=$xz
```

Comment:
note that according to the xz man page ```-T 0``` isn't implemented for decompression, but if it were you could add```extractopts=(-d -c -T 0)``` here to support that.